### PR TITLE
8327991: Improve HttpClient documentation with regard to reclaiming resources

### PR DESCRIPTION
**Pull request by dfuch**
Please find here a doc-only change that tries to better document how resources allocated by the HttpClient are reclaimed.
The `@implNote` in the HttpClient class level API documentation is extended to document what happens if you don't explicitly close the HttpClient. The note is refactored to introduce the notion of streaming or publishing bodies. BodySubscribers and BodyHandlers now have links that link back to the HttpClient API note, as well as identifying methods that return publishing or streaming body implementations.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change requires CSR request [JDK-8328100](https://bugs.openjdk.org/browse/JDK-8328100) to be approved
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issues
 * [JDK-8327991](https://bugs.openjdk.org/browse/JDK-8327991): Improve HttpClient documentation with regard to reclaiming resources (**Enhancement** - P3)
 * [JDK-8328100](https://bugs.openjdk.org/browse/JDK-8328100): Improve HttpClient documentation with regard to reclaiming resources (**CSR**)

### Reviewers
 * [Michael McMahon](https://openjdk.org/census#michaelm) (@Michael-Mc-Mahon - **Reviewer**) ⚠️ Review applies to [4f830897](https://git.openjdk.org/jdk/pull/18270/files/4f830897f8763c8e4641373b5d277df683adb878)
 * [Jaikiran Pai](https://openjdk.org/census#jpai) (@jaikiran - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18270/head:pull/18270` \
`$ git checkout pull/18270`

Update a local copy of the PR: \
`$ git checkout pull/18270` \
`$ git pull https://git.openjdk.org/jdk.git pull/18270/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18270`

View PR using the GUI difftool: \
`$ git pr show -t 18270`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18270.diff">https://git.openjdk.org/jdk/pull/18270.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18270#issuecomment-1994552543)

### CODE REVIEW

**Comment by dfuch** [comment]
A draft CSR can be seen here: https://bugs.openjdk.org/browse/JDK-8328100
Diff of the doc changes will be pasted into the CSR after this PR is reviewed. Let's agree on the wording here first.

File: src/java.net.http/share/classes/java/net/http/HttpClient.java [hunk 1]
```diff
@@ -132,21 +134,47 @@
  * reclaimed early by {@linkplain #close() closing} the client.
  *
  * @implNote
- *  <p id="closing">
- *  The JDK built-in implementation of the {@code HttpClient} overrides
+ *  <p id="streaming">
+ *  The HttpClient {@link BodyHandlers} and {@link BodySubscribers}
+ *  API provide some {@linkplain BodySubscribers##streaming-body streaming
```
Comment (Michael-Mc-Mahon) [inline src/java.net.http/share/classes/java/net/http/HttpClient.java:139]:
```suggestion
 *  The {@link BodyHandlers} and {@link BodySubscribers}
 *  classes provide some {@linkplain BodySubscribers##streaming-body streaming
```

File: src/java.net.http/share/classes/java/net/http/HttpResponse.java [hunk 2]
```diff
@@ -1199,8 +1240,8 @@ public static BodySubscriber<Path> ofFile(Path file) {
          * while blocking on read. In that case, the request will also be
          * cancelled and the {@code InputStream} will be closed.
          *
-         * @return a body subscriber that streams the response body as an
-         *         {@link InputStream}.
+         * @return a {@linkplain HttpClient##streaming streaming body subscriber}
+         *         that streams the response body as an {@link InputStream}.
```
Comment (Michael-Mc-Mahon) [inline src/java.net.http/share/classes/java/net/http/HttpResponse.java:1244]:
```suggestion
         *         which streams the response body as an {@link InputStream}.
```

File: src/java.net.http/share/classes/java/net/http/HttpResponse.java [hunk 3]
```diff
@@ -1225,8 +1266,8 @@ public static BodySubscriber<InputStream> ofInputStream() {
          * from being reused for subsequent operations.
          *
          * @param charset the character set to use when converting bytes to characters
-         * @return a body subscriber that streams the response body as a
-         *         {@link Stream Stream}{@code <String>}.
+         * @return a {@linkplain HttpClient##streaming streaming body subscriber} that streams
```
Comment (Michael-Mc-Mahon) [inline src/java.net.http/share/classes/java/net/http/HttpResponse.java:1269]:
```suggestion
         * @return a {@linkplain HttpClient##streaming streaming body subscriber} which streams
```

**Review by Michael-Mc-Mahon — APPROVED** [review]
LGTM

**Review by jaikiran — APPROVED** [review]
Hello Daniel, the changes look OK to me. package-info.java will need a copyright year update before integrating.

**Review by jaikiran — APPROVED** [review]

**Pull request closed without merging**